kornia/geometry/transform/homography_warper.py

```diff
@@ -164,6 +168,109 @@ def homography_warp3d(
     return F.grid_sample(patch_src, warped_grid, mode=mode, padding_mode=padding_mode, align_corners=align_corners)
 
 
+class Homography(nn.Module):
+    def __init__(self):
+        """Module to be used together with ImageRegistrator module for the optimization-based image
+        registration."""
+        super().__init__()
+        self.model = nn.Parameter(torch.Tensor(3, 3))
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        torch.nn.init.eye_(self.model)
+
+    def forward(self):
+        return torch.unsqueeze(self.model, dim=0)  # 1x3x3
+
+    def get_inverse_model(self):
+        return torch.unsqueeze(torch.inverse(self.model), dim=0)
+
+
+class ImageRegistrator(nn.Module):
+    """'Module, which performs optimization-based image registration, similar to https://github.com/kornia/kornia-
+    examples/blob/master/homography.ipynb."""
+    def __init__(self,
+                 model_type: str = 'H',
+                 optimizer: torch.optim.Optimizer = optim.Adam,
+                 loss_fn = F.l1_loss,
+                 pyr_levels = 5,
+                 lr: float = 1e-3,
+                 n_iter: int = 100):
+        super().__init__()
+        self.model_type = model_type
+        if model_type == "H":
+            self.warper = HomographyWarper
+            self.model = Homography()
+        else:
+            raise ValueError(f"{model_type} is not supported. Try 'H'")
+        self.pyr_levels = pyr_levels
+        self.optimizer = optimizer
+        self.lr = lr
+        self.loss_fn = loss_fn
+        self.n_iter = n_iter
+
+    def get_single_level_loss(self,
+                              img_src: torch.Tensor,
+                              img_dst: torch.Tensor,
+                              transform_model: torch.nn.Module) -> torch.Tensor:
+        # ToDo: Make possible registration of images of different shape
+        assert len(img_src.shape) == len(img_dst.shape), (img_src.shape, img_dst.shape)
+        _height, _width = img_src.shape[-2:]
+        warper = self.warper(_height, _width)
+        img_src_to_dst = warper(img_src, transform_model)
+        # compute and mask loss
+        loss = self.loss_fn(img_src_to_dst, img_dst, reduction='none') # 1x3xHxW
+        ones = warper(torch.ones_like(img_src), transform_model)
+        loss = loss.masked_select(ones > 0.9).mean()
+        return loss
+
+    def reset_model(self):
+        self.model.reset_parameters()
+
+    def register(self,
+                 src_img: torch.Tensor,
+                 dst_img: torch.Tensor,
+                 verbose: bool = False,
+                 device=torch.device('cpu')) -> torch.Tensor:
+        '''Example:
+        >>>> registrator = ImageRegistration('H', lr=1e-3, n_iter=50, pyr_levels=5,
+                               loss_fn = F.l1_loss)
+        >>>> homo = registrator.register(img_src, img_dst, True)
+        '''
+        # compute the gaussian pyramids
+        self.to(device)
```

Comment:
probably we should remove this outside like in regular nets and its the user who sends to the appropriate device